src/Tools/dotnet-monitor/Egress/S3/MultiPartUploadStream.cs

```diff
@@ -47,13 +59,43 @@ public override async Task FlushAsync(CancellationToken cancellationToken)
 
     public async Task FinalizeAsync(CancellationToken cancellationToken)
     {
+        _finalize = true;
+
+        if (_writeSynchronousArtifacts != null)
+            await _writeSynchronousArtifacts;
+        
         if (Closed)
             throw new ObjectDisposedException(nameof(MultiPartUploadStream));
         if (_offset == 0)
             return;
         await DoWriteAsync(true, cancellationToken);
     }
 
+    public async Task WriteAsyncLoop(CancellationToken cancellationToken)
```

Comment:
This method appears to constantly loop (with the delay) until data is put into the buffer.

What it should do is wait on a semaphore that represents when data is available, await that, lock on a sync object to read the data out of the buffer into a temporary buffer, release the lock, then upload the temporary buffer data via the WriteAsync method.

The Write method should acquire the lock on the sync object, write the data to the buffer, release the lock, and release the data available semaphore.

By doing this, the WriteAsyncLoop method avoids spinning constantly and only does work when data is available.